### PR TITLE
fixes #7596 - validate plugin dependencies when loading, various fixes

### PR DESCRIPTION
**Pull request by domcleal**

### CODE REVIEW

File: test/plugin_test.rb [hunk 1]
```diff
@@ -20,4 +20,51 @@ def test_http_rackup_returns_empty_string_with_missing_rackup_path
...
+    end
+  end
+
+  def test_unsatisified_dependency
+    eval "class TestPlugin5 < Proxy::Plugin; plugin :test5, '1.5'; end"
+    eval "class TestPlugin6 < Proxy::Plugin; plugin :test6, '1.0'; requires :test5, '> 2.0'; end"
+    assert_nothing_raised do
+      TestPlugin5.new.validate_dependencies!(TestPlugin5.dependencies)
+    end
+    assert_raise(::Proxy::PluginVersionMismatch) do
```
Comment (dmitri-d) [inline test/plugin_test.rb:39]: this bit is quite similar to `test_core_dependency`. Otoh, there's no test that verifies that a plugin with satisfiable dependency is valid...

**Comment by dmitri-d** [comment]
looks good, just the small issue in tests.

File: test/plugin_test.rb (hunk 1 above)
Comment (domcleal) [inline test/plugin_test.rb:39]: Yeah, I've cleared it up now into two separate tests, one for satisfied deps and one for unsatisfied.  core_dependency was meant to test the case with a -develop suffix, so I've added a comment above.

**Comment by domcleal** [comment]
I had to change the tests from using eval to simply declaring the classes after a very strange issue.  When using find_plugin, I'd get back a different (unnamespaced) class object to the one I'd get when just using the constant in the test.

Updated.

**Comment by dmitri-d** [comment]
merged in at 20b9f96. thanks, @domcleal!

**Pull request closed without merging**